core/src/main/scala/org/apache/spark/util/collection/Utils.scala

```diff
@@ -42,6 +42,23 @@ private[spark] object Utils extends SparkCollectionUtils {
     ordering.leastOf(input.asJava, num).iterator.asScala
   }
 
+  /**
+   * Returns the last K elements from the input.
+   */
+  def takeLast[T](input: Iterator[T], num: Int): Iterator[T] = {
+    assert(num >= 0)
+    if (input.isEmpty || num == 0) {
+      Iterator.empty[T]
+    } else {
+      var last = Seq.empty[T]
+      val grouped = input.grouped(num)
```

Comment:
I am not sure whether we should use externalsorter to reverse the partition instead of this grouped iterator